BSP/STM32F413/Src/BSP_CAN.c

```diff
@@ -122,26 +198,63 @@ uint8_t BSP_CAN_Write(uint32_t id, uint8_t data[8], uint8_t length) {
  * @pre     The data parameter must be at least 8 bytes or hardfault may occur.
  * @param   id : pointer to store id of the message that was received.
  * @param   data : pointer to store data that was received. Must be 8bytes or bigger.
- * @return  0 if nothing was received so ignore id and data that was received. Any other value indicates data was received and stored.
+ * @return  ERROR if nothing was received so ignore id and data that was received. SUCCESS indicates data was received and stored.
  */
-uint8_t BSP_CAN_Read(uint32_t *id, uint8_t *data) {
-    if(RxFlag){
-		for(int i = 0; i < 8; i++){
-			data[i] = RxMessage.Data[i];
-		}
-        *id = RxMessage.StdId;
-		RxFlag = false;
-		return 1;
-	}
-    return 0;
+ErrorStatus BSP_CAN_Read(uint32_t *id, uint8_t *data) {
+    // If the queue is empty, return err
+    if(gRxQueue.head == gRxQueue.tail) {
+        return ERROR;
+    }
+    
+    // Get the message
+    msg_t msg;
+    QueueGet(&gRxQueue, &msg);
+
+    // Transfer the message to the provided pointers
+    for(int i = 0; i < 8; i++){
+        data[i] = msg.data[i];
+    }
+    *id = msg.id;
+
+    return SUCCESS;
 }
 
-void CAN1_RX0_IRQHandler(void)
-{
-    CAN_Receive(CAN1, CAN_FIFO0, &RxMessage);
+void CAN1_RX0_IRQHandler(void) {
+    asm("CPSID I");   // Disable Interrupts
+    OSIntEnter();     // Signal to uC/OS
```

Comment:
https://doc.micrium.com/display/osiiidoc/Typical+uC-OS-III+Interrupt+Service+Routine
I used this page to make my ISR. I don't believe steps 5 and 6 are implemented. I did something similar in the I2C ISR for reference.